File: springboot-rest-scala-jpa-contacts/src/test/scala/com/tnsilver/contacts/repository/ContactRepositoryIT.scala
/*
 * File: scala
 * Creation Date: Jun 25, 2021
 *
 * Copyright (c) 2021 T.N.Silverman - all rights reserved
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses  this file to you under the Apache License, Version
 * 2.0 (the "License");  you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tnsilver.contacts.repository

import com.tnsilver.contacts.base.BaseJpaTest
import com.tnsilver.contacts.model.{Contact, SocialSecurityNumber}
import org.junit.jupiter.api.Assertions._
import org.junit.jupiter.api._
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.data.domain.Sort.Direction
import org.springframework.data.domain.{Page, PageRequest, Pageable, Sort}
import org.springframework.transaction.annotation.Transactional

import java.time.LocalDate
import javax.annotation.Resource
/**
 * The test JpaContactRepositoryIT tests the {@link ContactRepository} functionality
 *
 * @author T.N.Silverman
 *
 */
@SpringBootTest
@Transactional class ContactRepositoryIT extends BaseJpaTest {

  private val PAGEABLE : Pageable = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"))

  @Resource private val contactRepository : ContactRepository = null

  @BeforeAll
  @throws[Exception]
  override def beforeAll(): Unit = {
    super.beforeAll()
  }

  @BeforeEach
  @throws[Exception]
  override def beforeEach(info: TestInfo): Unit = {
    super.beforeEach(info)
  }

  @ParameterizedTest
  @DisplayName("test get filtered contacts")
  @CsvSource(Array("123-45-6789",
    "123-45-7890",
    "123-45-8901",
    "123-45-9012",
    "123-45-1012"))
  def whenGetFilteredContacts_ThenNotEmpty(ssn: String): Unit = {
    val filter : Contact = new Contact(new SocialSecurityNumber(ssn), null, null, null, null, null)
    val actual : Page[Contact] = contactRepository.getFilteredContacts(filter, PAGEABLE)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
  }

  @ParameterizedTest
  @DisplayName("test get filtered contacts with first name")
  @CsvSource(Array("123-45-6789,H",
    "123-45-7890,Marge",
    "123-45-8901,Lis",
    "123-45-9012,Ba",
    "123-45-1012,Maggi"))
  def whenGetFilteredContactsWithFirstName_ThenNotEmpty(ssn: String, firstName: String): Unit = {
    val filter = new Contact(new SocialSecurityNumber(ssn), firstName, null, null, null, null)
    val actual : Page[Contact] = contactRepository.getFilteredContacts(filter, PAGEABLE)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
  }

  @ParameterizedTest
  @DisplayName("test get filtered contacts with names")
  @CsvSource(Array("123-45-6789,H,Simp",
    "123-45-7890,Marge,Si",
    "123-45-8901,Lis,Simpson",
    "123-45-9012,Ba,Sim",
    "123-45-1012,Maggie,S"))
  def whenGetFilteredContactsWithName_ThenNotEmpty(ssn: String, firstName: String, lastName: String): Unit = {
    val filter = new Contact(new SocialSecurityNumber(ssn), firstName, lastName, null, null, null)
    val actual = contactRepository.getFilteredContacts(filter, PAGEABLE)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
  }

  @ParameterizedTest
  @DisplayName("test get filtered contacts with birth date") // @formatter:off
  @CsvSource(Array("123-45-6789,H,Simp,1978-06-20",
    "123-45-7890,Marge,Si,1979-06-20",
    "123-45-8901,Lis,Simpson,1993-01-13",
    "123-45-9012,Ba,Sim,1995-09-18",
    "123-45-1012,Maggie,S,2003-03-05"))
  def whenGetFilteredContactsWithBirthDate_ThenNotEmpty(ssn: String, firstName: String, lastName: String, birthDate: String): Unit = {
    val filter = new Contact(new SocialSecurityNumber(ssn), firstName, lastName, LocalDate.parse(birthDate), null, null)
    val actual : Page[Contact] = contactRepository.getFilteredContacts(filter, PAGEABLE)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
  }

  @Test def testGetFilteredContactsByMarried(): Unit = {
    val filter : Contact = new Contact(new SocialSecurityNumber("123-45-6789"), "Home", "Simp", LocalDate.parse("1978-06-20"), true, null)
    val pageable : Pageable = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"))
    val actual : Page[Contact] = contactRepository.getFilteredContacts(filter, pageable)
    assertFalse(actual.getContent.isEmpty)
  }

  @Test def testGetFilteredContactsByChildren(): Unit = {
    val filter = new Contact(new SocialSecurityNumber("123-45-6789"), "Home", "Simp", LocalDate.parse("1978-06-20"), true, 3)
    val pageable = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getFilteredContacts(filter, pageable)
    assertFalse(actual.getContent.isEmpty)
  }

  @Test def testGetFilteredContactsByNothing(): Unit = {
    val filter = new Contact
    val pageable = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getFilteredContacts(filter, pageable)
    assertFalse(actual.getContent.isEmpty)
  }

  @ParameterizedTest
  @DisplayName("test get by params") // @Disabled
  // @formatter:off
  @CsvSource(value = Array("123-45-6789,'','','','',''",
    "'','','','','',3",
    "'','','',1978-06-20,'',''",
    "'','','',1978-06-20,true,3",
    "'','',Simp,1978-06-20,true,3",
    "'',Hom,Simp,1978-06-20,true,3",
    "123-45-6789,H,S,1978-06-20,true,3",
    "123-45-6789,H,'',1978-06-20,true,3"))
  // @formatter:on
  def testGetByParams(_ssn: String, firstName: String, lastName: String, _birthDate: String, _married: String, _children: String): Unit = {
    val ssn = if (null == _ssn || _ssn.isBlank) null
    else new SocialSecurityNumber(_ssn)
    val birthDate = if (null == _birthDate || _birthDate.isBlank) null
    else LocalDate.parse(_birthDate)
    val married : java.lang.Boolean = if (null == _married || _married.isBlank) null else java.lang.Boolean.valueOf(_married)
    val children = if (null == _children || _children.isBlank) null
    else Integer.valueOf(_children)
    val filter : Contact = new Contact(ssn, firstName, lastName, birthDate, married, children)
    val pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getByParams(if (filter.getSsn == null) null
    else filter.getSsn.getSsn, filter.getFirstName, filter.getLastName, filter.getBirthDate, filter.getMarried, filter.getChildren, pageable)
    assertFalse(actual.getContent.isEmpty)
    System.out.println()
    logger.info("{} results:", actual.getContent.size)
    actual.getContent.forEach((c: Contact) => logger.info("{}", c))
    System.out.println("")
  }

  @Test def testGetBySsn(): Unit = {
    val ssn = "123-45-6789"
    val pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getByParams(ssn, null, null, null, null, null, pageable)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
    System.out.println()
    logger.info("{} results:", actual.getContent.size)
    actual.getContent.forEach((c: Contact) => logger.info("{}", c))
    System.out.println()
  }

  @Test def testGetByFirstName(): Unit = {
    val firstName = "Hom"
    val pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getByParams(null, firstName, null, null, null, null, pageable)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
    System.out.println()
    logger.info("{} results:", actual.getContent.size)
    actual.getContent.forEach((c: Contact) => logger.info("{}", c))
    System.out.println()
  }

  @Test def testGetBySSnAndFirstName(): Unit = {
    val ssn = "123-45-6789"
    val firstName = "Hom"
    val pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getByParams(ssn, firstName, null, null, null, null, pageable)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
    System.out.println()
    logger.info("{} results:", actual.getContent.size)
    actual.getContent.forEach((c: Contact) => logger.info("{}", c))
    System.out.println()
  }

  @Test def testGetBySSnAndNames(): Unit = {
    val ssn = "123-45-6789"
    val firstName = "Hom"
    val lastName = "Simp"
    val pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getByParams(ssn, firstName, lastName, null, null, null, pageable)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
    System.out.println()
    logger.info("{} results:", actual.getContent.size)
    actual.getContent.forEach((c: Contact) => logger.info("{}", c))
    System.out.println()
  }

  @Test def testGetBySSnAndNamesAndMarried(): Unit = {
    val ssn = "123-45-6789"
    val firstName = "Hom"
    val lastName = "Simp"
    val married = null
    val pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getByParams(ssn, firstName, lastName, null, married, null, pageable)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
    System.out.println()
    logger.info("{} results:", actual.getContent.size)
    actual.getContent.forEach((c: Contact) => logger.info("{}", c))
    System.out.println()
  }

  @Test def testGetBySSnAndNamesAndMarriedAndChildren(): Unit = {
    val ssn = "123-45-6789"
    val firstName = "Hom"
    val lastName = "Simp"
    val married = true
    val children = Integer.valueOf(3)
    val pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getByParams(ssn, firstName, lastName, null, married, children, pageable)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
    System.out.println()
    logger.info("{} results:", actual.getContent.size)
    actual.getContent.forEach((c: Contact) => logger.info("{}", c))
    System.out.println()
  }

  @Test def testGetBirthDate(): Unit = {
    val birthDate = LocalDate.parse("1978-06-20")
    val pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"))
    val actual = contactRepository.getByParams(null, null, null, birthDate, null, null, pageable)
    assertFalse(actual.getContent.isEmpty)
    assertEquals(1, actual.getContent.size)
    System.out.println()
    logger.info("{} results:", actual.getContent.size)
    actual.getContent.forEach((c: Contact) => logger.info("{}", c))
    System.out.println()
  }
}

File: springboot-rest-jpa-contacts/src/test/java/com/tnsilver/contacts/repository/ContactRepositoryIT.java
/*
 * File: ContactRepositoryIT.java
 * Creation Date: Jul 20, 2021
 *
 * Copyright (c) 2021 T.N.Silverman - all rights reserved
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses  this file to you under the Apache License, Version
 * 2.0 (the "License");  you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tnsilver.contacts.repository;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.time.LocalDate;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.domain.Sort.Direction;
import org.springframework.transaction.annotation.Transactional;

import com.tnsilver.contacts.base.BaseJpaTest;
import com.tnsilver.contacts.model.Contact;
import com.tnsilver.contacts.model.SocialSecurityNumber;

import lombok.extern.slf4j.Slf4j;

/**
 * The test JpaContactRepositoryIT tests the {@link ContactRepository} functionality
 *
 * @author T.N.Silverman
 *
 */
@SpringBootTest
@Transactional
@Slf4j
public class ContactRepositoryIT extends BaseJpaTest {

    @Autowired private ContactRepository contactRepository;
    private static final Pageable PAGEABLE = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"));

    @BeforeAll
    public static void beforeAll() throws Exception {
        BaseJpaTest.beforeAll();
    }

    @BeforeEach
    public void beforeEach(TestInfo info) throws Exception {
        super.beforeEach(info);
    }

    @ParameterizedTest
    @DisplayName("test filter contacts by ssn")
    // @formatter:off
    @CsvSource({ "123-45-6789,Homer",
                 "123-45-7890,Marge",
                 "123-45-8901,Lisa",
                 "123-45-9012,Bart",
                 "123-45-1012,Maggie" })
    // @formatter:on
    public void whenGetFilterBySsn_ThenFindExpected(String ssn, String expected) {
        Contact filter = new Contact(new SocialSecurityNumber(ssn), null, null, null, null, null);
        Page<Contact> actual = contactRepository.getFilteredContacts(filter, PAGEABLE);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        assertEquals(expected, actual.getContent().get(0).getFirstName());
    }

    @ParameterizedTest
    @DisplayName("test filtered contacts by partial first name")
    // @formatter:off
    @CsvSource({ "123-45-6789,H",
                 "123-45-7890,Marge",
                 "123-45-8901,Lis",
                 "123-45-9012,Ba",
                 "123-45-1012,Maggi" })
    // @formatter:on
    public void whenFilterByFirstName_ThenFindExpected(String ssn, String firstName) {
        Contact filter = new Contact(new SocialSecurityNumber(ssn), firstName, null, null, null, null);
        Page<Contact> actual = contactRepository.getFilteredContacts(filter, PAGEABLE);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        assertTrue(actual.getContent().get(0).getFirstName().startsWith(firstName));
    }

    @ParameterizedTest
    @DisplayName("test filter contacts by names")
    // @formatter:off
    @CsvSource({ "123-45-6789,H,Simp",
                 "123-45-7890,Marge,Si",
                 "123-45-8901,Lis,Simpson",
                 "123-45-9012,Ba,Sim",
                 "123-45-1012,Maggie,S" })
    // @formatter:on
    public void whenFilterByNames_ThenFindExpected(String ssn, String firstName, String lastName) {
        Contact filter = new Contact(new SocialSecurityNumber(ssn), firstName, lastName, null, null, null);
        Page<Contact> actual = contactRepository.getFilteredContacts(filter, PAGEABLE);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        assertTrue(actual.getContent().get(0).getFirstName().startsWith(firstName));
        assertTrue(actual.getContent().get(0).getLastName().startsWith(lastName));
    }

    @ParameterizedTest
    @DisplayName("test filter by birth date")
    // @formatter:off
    @CsvSource({ "123-45-6789,H,Simp,1978-06-20",
                 "123-45-7890,Marge,Si,1979-06-20",
                 "123-45-8901,Lis,Simpson,1993-01-13",
                 "123-45-9012,Ba,Sim,1995-09-18",
                 "123-45-1012,Maggie,S,2003-03-05" })
    // @formatter:off
    public void whenFilterByBirthDate_ThenFindExpected(String ssn,String firstName,String lastName,String birthDate) {
        Contact filter = new Contact(new SocialSecurityNumber(ssn),firstName, lastName,LocalDate.parse(birthDate), null, null);
        Page<Contact> actual = contactRepository.getFilteredContacts(filter, PAGEABLE);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        assertTrue(actual.getContent().get(0).getFirstName().startsWith(firstName));
        assertTrue(actual.getContent().get(0).getLastName().startsWith(lastName));
        assertTrue(actual.getContent().get(0).getBirthDate().equals(filter.getBirthDate()));
    }

    @Test
    @DisplayName("test filter by married")
    public void whenFilterByMarried_ThenFindExpected() {
        Contact filter = new Contact(new SocialSecurityNumber("123-45-6789"), "Home", "Simp", LocalDate.parse("1978-06-20"), true, null);
        Pageable pageable = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getFilteredContacts(filter, pageable);
        assertFalse(actual.getContent().isEmpty());
        assertTrue(actual.getContent().get(0).getFirstName().startsWith(filter.getFirstName()));
        assertTrue(actual.getContent().get(0).getLastName().startsWith(filter.getLastName()));
        assertTrue(actual.getContent().get(0).getBirthDate().equals(filter.getBirthDate()));
    }

    @Test
    @DisplayName("test filter by last name and married")
    public void whenFilterByLastNameAndMarried_ThenFindExpected() {
        Contact filter = new Contact(null, "", "Simp", null, true, null);
        Pageable pageable = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getFilteredContacts(filter, pageable);
        assertFalse(actual.getContent().isEmpty());
        // homer and marge (married)
        assertEquals(2, actual.getContent().size());
        actual.getContent().forEach(c -> assertTrue(c.getLastName().startsWith("Simp")));
    }


    @Test
    @DisplayName("test filter by children")
    public void whenFilterByChildren_ThenFindExpected() {
        Contact filter = new Contact(new SocialSecurityNumber("123-45-6789"), "Home", "Simp", LocalDate.parse("1978-06-20"), true, 3);
        Pageable pageable = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getFilteredContacts(filter, pageable);
        assertFalse(actual.getContent().isEmpty());
        assertTrue(actual.getContent().get(0).getFirstName().startsWith(filter.getFirstName()));
        assertTrue(actual.getContent().get(0).getLastName().startsWith(filter.getLastName()));
        assertTrue(actual.getContent().get(0).getBirthDate().equals(filter.getBirthDate()));
        assertTrue(actual.getContent().get(0).getMarried().equals(filter.getMarried()));
        assertTrue(actual.getContent().get(0).getChildren().equals(filter.getChildren()));
    }

    @Test
    @DisplayName("test filter by nothing")
    public void whenFilterNothing_ThenFindAll() {
        Contact filter = new Contact();
        filter.setId(null);
        filter.setSsn(null);
        Pageable pageable = PageRequest.of(0, 10, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getFilteredContacts(filter, pageable);
        assertFalse(actual.getContent().isEmpty());
    }

    @ParameterizedTest
    @DisplayName("test get by params")
    // @Disabled
    // @formatter:off
     @CsvSource(value = { "123-45-6789,'','','','',''",
                          "'','','','','',3",
                          "'','','',1978-06-20,'',''",
                          "'','','',1978-06-20,true,3",
                          "'','',Simp,1978-06-20,true,3",
                          "'',Hom,Simp,1978-06-20,true,3",
                          "123-45-6789,H,S,1978-06-20,true,3",
                          "123-45-6789,H,'',1978-06-20,true,3" })
    // @formatter:on
    public void whenGetByParams_ThenFound(String _ssn, String firstName, String lastName, String _birthDate,
        String _married, String _children) {
        SocialSecurityNumber ssn = null == _ssn || _ssn.isBlank() ? null : new SocialSecurityNumber(_ssn);
        LocalDate birthDate = null == _birthDate || _birthDate.isBlank() ? null : LocalDate.parse(_birthDate);
        Boolean married = null == _married || _married.isBlank() ? null : Boolean.valueOf(_married);
        Integer children = null == _children || _children.isBlank() ? null : Integer.valueOf(_children);
        Contact filter = new Contact(ssn, firstName, lastName, birthDate, married, children);
        PageRequest pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getByParams(filter.getSsn() == null ? null : filter.getSsn().getSsn(),
                                                             filter.getFirstName(), filter.getLastName(),
                                                             filter.getBirthDate(), filter.getMarried(),
                                                             filter.getChildren(), pageable);
        assertFalse(actual.getContent().isEmpty());
        System.out.println();
        log.debug("{} results:", actual.getContent().size());
        actual.getContent().forEach(c -> log.debug("{}", c));
        System.out.println();
    }

    @Test
    @DisplayName("test get by ssn param")
    public void whenGetBySsnParam_ThenFound() {
        String ssn = "123-45-6789";
        PageRequest pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getByParams(ssn, null, null, null, null, null, pageable);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        System.out.println();
        log.debug("{} results:", actual.getContent().size());
        actual.getContent().forEach(c -> log.debug("{}", c));
        System.out.println();
    }

    @Test
    @DisplayName("test get by first name param")
    public void whenGetByFirstNameParam_ThenFound() {
        String firstName = "Hom";
        PageRequest pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getByParams(null, firstName, null, null, null, null, pageable);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        System.out.println();
        log.debug("{} results:", actual.getContent().size());
        actual.getContent().forEach(c -> log.debug("{}", c));
        System.out.println();
    }

    @Test
    @DisplayName("test get by names params")
    public void whenGetByNamesParams_ThenFound() {
        String ssn = "123-45-6789";
        String firstName = "Hom";
        PageRequest pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getByParams(ssn, firstName, null, null, null, null, pageable);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        System.out.println();
        log.debug("{} results:", actual.getContent().size());
        actual.getContent().forEach(c -> log.debug("{}", c));
        System.out.println();
    }

    @Test
    @DisplayName("test get by ssn and names params")
    public void whenGetBySsnAndNamesParams_ThenFound() {
        String ssn = "123-45-6789";
        String firstName = "Hom";
        String lastName = "Simp";
        PageRequest pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getByParams(ssn, firstName, lastName, null, null, null, pageable);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        System.out.println();
        log.debug("{} results:", actual.getContent().size());
        actual.getContent().forEach(c -> log.debug("{}", c));
        System.out.println();
    }

    @Test
    @DisplayName("test get by ssn, names, married params")
    public void whenGetBySsnNamesAndMarriedParams_ThenFound() {
        String ssn = "123-45-6789";
        String firstName = "Hom";
        String lastName = "Simp";
        Boolean married = null;
        PageRequest pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getByParams(ssn, firstName, lastName, null, married, null, pageable);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        System.out.println();
        log.debug("{} results:", actual.getContent().size());
        actual.getContent().forEach(c -> log.debug("{}", c));
        System.out.println();
    }

    @Test
    @DisplayName("test get by ssn, names, married, children params")
    public void whenGetBySsnNamesMarriedAndChildrenParams_ThenFound() {
        String ssn = "123-45-6789";
        String firstName = "Hom";
        String lastName = "Simp";
        Boolean married = true;
        Integer children = Integer.valueOf(3);
        PageRequest pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getByParams(ssn, firstName, lastName, null, married, children,
                                                             pageable);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        System.out.println();
        log.debug("{} results:", actual.getContent().size());
        actual.getContent().forEach(c -> log.debug("{}", c));
        System.out.println();
    }

    @Test
    @DisplayName("test get by birthdate param")
    public void whenGetByBirthdateParam_ThenFound() {
        LocalDate birthDate = LocalDate.parse("1978-06-20");
        PageRequest pageable = PageRequest.of(0, 1000, Sort.by(Direction.ASC, "id"));
        Page<Contact> actual = contactRepository.getByParams(null, null, null, birthDate, null, null, pageable);
        assertFalse(actual.getContent().isEmpty());
        assertEquals(1, actual.getContent().size());
        System.out.println();
        log.debug("{} results:", actual.getContent().size());
        actual.getContent().forEach(c -> log.debug("{}", c));
        System.out.println();
    }
}
